src/dataflow/src/render/mod.rs

```diff
@@ -692,6 +706,38 @@ where
     }
 }
 
+/// A trait meant to capture requisite functionality for use as a timestamp.
+pub trait SystemEventTime: Timestamp {
+    /// A reference to the system time component of the timestamp.
+    fn system_time(&mut self) -> &mut repr::Timestamp;
+    /// An additive delay to system time.
+    fn system_delay(delay: repr::Timestamp) -> <Self as Timestamp>::Summary;
+    /// A reference to the event time component of the timestamp.
+    fn event_time(&mut self) -> &mut repr::Timestamp;
+    /// An additive delay to event time.
+    fn event_delay(delay: repr::Timestamp) -> <Self as Timestamp>::Summary;
+    /// Move the timestamp backwards.
```

Comment:
super surprising to see `retreat` here because I feel like "timestamps can't go backwards" has been so drilled into me. would be nice to maybe say a few words about where this should / shouldn't be used

Reply:
I swear I typed a response to this, but it seems to have been lost to time.  This is playing the same role as `Alt`/`Neu` which are no longer used here. It has the same role as `saturating_sub(1)`, which we sort of use in the coordinator too. It gets used to provide enough of a barrier that logical compaction doesn't conflate two distinct times.